pkg/kube/object_patch/patch.go

```diff
@@ -0,0 +1,344 @@
+package object_patch
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"gopkg.in/yaml.v3"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/flant/shell-operator/pkg/app"
+	"github.com/flant/shell-operator/pkg/jq"
+	"github.com/flant/shell-operator/pkg/kube"
+)
+
+type ObjectPatcher struct {
+	kubeClient kube.KubernetesClient
+}
+
+func NewObjectPatcher(kubeClient kube.KubernetesClient) *ObjectPatcher {
+	return &ObjectPatcher{kubeClient: kubeClient}
+}
+
+func (o *ObjectPatcher) ParseSpecs(specBytes []byte) ([]OperationSpec, error) {
+	specs, err := unmarshalFromJsonOrYaml(specBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	err = validateSpecs(specs)
+	if err != nil {
+		return nil, err
+	}
+
+	return specs, err
+}
+
+func (o *ObjectPatcher) GenerateFromJSONAndExecuteOperations(specs []OperationSpec) error {
+	var applyErrors *multierror.Error
+	for _, spec := range specs {
+		var operationError error
+
+		switch spec.Operation {
+		case Create:
+			operationError = o.CreateObject(&unstructured.Unstructured{Object: spec.Object}, spec.Namespace, spec.Subresource)
+		case CreateOrUpdate:
+			operationError = o.CreateOrUpdateObject(&unstructured.Unstructured{Object: spec.Object}, spec.Namespace, spec.Subresource)
+		case Delete:
+			operationError = o.DeleteObject(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name)
+		case DeleteInBackground:
+			operationError = o.DeleteObjectInBackground(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name)
+		case DeleteNonCascading:
+			operationError = o.DeleteObjectNonCascading(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name)
+		case JQPatch:
+			operationError = o.JQPatchObject(spec.JQFilter, spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name, spec.Subresource)
+		case MergePatch:
+			jsonMergePatch, err := json.Marshal(spec.MergePatch)
+			if err != nil {
+				applyErrors = multierror.Append(applyErrors, err)
+			}
+
+			operationError = o.MergePatchObject(jsonMergePatch, spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name, spec.Subresource)
+		case JSONPatch:
+			jsonJsonPatch, err := json.Marshal(spec.MergePatch)
+			if err != nil {
+				applyErrors = multierror.Append(applyErrors, err)
+			}
+
+			operationError = o.JSONPatchObject(jsonJsonPatch, spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name, spec.Subresource)
+		}
+
+		if operationError != nil {
+			applyErrors = multierror.Append(applyErrors, operationError)
+		}
+	}
+
+	//goland:noinspection GoNilness
+	return applyErrors.ErrorOrNil()
+}
+
+func unmarshalFromJsonOrYaml(specs []byte) ([]OperationSpec, error) {
```

Comment:
Do we need to support YAML responses from hooks?

Reply:
Why not? Much easier to write by hand in simple hooks.